walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,236 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * Finds a list of potential meeting times based on the meeting request and a list of existing
+   * events and their attendees. Picks time slots if one or more time slots exists so that both
+   * mandatory and optional attendees can attend. Otherwise, returns the time slots that fit just
+   * the mandatory attendees.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long meetingDurationMinutes = request.getDuration();
+
+    List<TimeRange> mandatoryAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> optionalAttendeesBusyTimeRanges =
+        getAllBusyTimeRanges(events, optionalAttendees);
+
+    sortTimeRangesByStart(mandatoryAttendeesBusyTimeRanges);
+    sortTimeRangesByStart(optionalAttendeesBusyTimeRanges);
+
+    List<TimeRange> mandatoryAttendeesFreeTimeRanges =
+        getFreeTimeRanges(mandatoryAttendeesBusyTimeRanges, meetingDurationMinutes);
+    List<TimeRange> optionalAttendeesFreeTimeRanges =
+        getFreeTimeRanges(optionalAttendeesBusyTimeRanges, meetingDurationMinutes);
+
+    if (mandatoryAttendees.isEmpty()) {
+      return optionalAttendeesFreeTimeRanges;
+    } else if (optionalAttendees.isEmpty()) {
+      return mandatoryAttendeesFreeTimeRanges;
+    } else {
+      return findCommonFreeTimeRanges(
+          mandatoryAttendeesFreeTimeRanges,
+          optionalAttendeesFreeTimeRanges,
+          meetingDurationMinutes);
+    }
+  }
+
+  /**
+   * Constructs all the time ranges that any of the mandatory attendees is busy. These time ranges
+   * cannot be part of the potential meeting time.
+   */
+  private List<TimeRange> getAllBusyTimeRanges(
+      Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
```

Comment:
Since you're going to update this `busyTimes` variable to an entirely new `List` on each iteration, this is a good opportunity to use `ImmutableList` from Guava if you like. You can find more about it (including how to add the dependency to your project) here https://github.com/google/guava.

You don't have to do it though for this if you don't want to, but if you're interested you can.

Reply:
I wonder if busyTimes in this method is immutable? Based on my implementation right now, I update busyTimes to the output of `getUpdatedBusyTimeRanges(currEventTimeRange, busyTimes);` -- wouldn't this make busyTimes not immutable?

Reply:
The variable `busyTimes` is mutable (not `final`) but the value that it points to is immutable because you always replace it with a completely new copy rather than updating the existing value in place. We can make a distinction between the mutability of the variable and the value that it points to:
```
List<String> foo = new ArrayList<>(); // Both variable and value mutable
final List<String> foo = new ArrayList<>(); // Variable immutable and value mutable
List<String> foo = ImmutableList.of(); // Variable mutable, value immutable
final List<String> foo = ImmutableList.of(); // Both immutable
```

In your case, I think you could define `ImmutableList<String> busyTimes` and keep updating the mutable variable with a new immutable list at each iteration.